active/003-seq-feature.md

```diff
@@ -0,0 +1,98 @@
+---
+Feature name: interval-metadata
+Start date: <2015-12-18>
+Pull request: 1
+Authors: ["@RNAer", "@mortonjt"]
+Contributors: ["@RNAer", "@mortonjt", "@ebolyen", "@jairideout", "@wasade", "@gregcaporaso", "@rob-knight"]
+---
+
+# Summary
+
+Introduce an `interval_metadata` object to allow for the storage and querying of intervals.
+
+# Motivation
+
+The current implementation of `positional_metadata` in the `Sequence` class takes > 1TB memory to read in a E. coli genome with 260K features. And the `pandas` sparse data frame can alleviate the mem usage, but it is still too memory hogging. The discussion is mainly in this [thread](https://github.com/biocore/scikit-bio/issues/1159).  Storing intervals instead of `positional_metadata` would substantially save on memory.
+
+Additional benefits of having a `interval_metadata` objects include the ability to rapidly query features by interval.  IntervalTrees allow for both querying by position and ranges to retrieve all overlapping intervals.  This can be particularly helpful when dealing with layers of features (i.e. transcripts made up of exons, or operons made up of genes).
+
+# Detailed design
+
+We propose to use the `IntervalTree` object in `bx-python.`  Some benchmarks have been made for various IntervalTree data structures - in this [benchmark](https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was determined to have one of the fastest query times.
+
+## `IntervalMetadata` object structure
+features : dictionary.  keys are `skbio.Feature` objects.  values are a list of tuples, where each tuple corresponds to an interval.
+intervals : bx-python `IntervalTree` object.  The keys correspond to intervals and the values correspond to a single `skbio.Feature` object.   
+
+These objects are arranged such that features can be queried by both feature attributes and intervals.
+
+## `Feature` object structure
+The feature object is more a less a immutable, hashable dictionary.  This store arbiturary information about features.  For instance, strand information, CDS, ...
```

Comment:
Thanks. See that now. The joys of github comments combined with email

On Fri, Apr 29, 2016 at 11:07 AM, Zech Xu <notifications@github.com> wrote:

> In active/003-seq-feature.md
> <https://github.com/biocore/scikit-bio-rfcs/pull/9#discussion_r61621223>:
>
> > +The current implementation of `positional_metadata` in the `Sequence` class takes > 1TB memory to read in a E. coli genome with 260K features. And the `pandas` sparse data frame can alleviate the mem usage, but it is still too memory hogging. The discussion is mainly in this [thread](https://github.com/biocore/scikit-bio/issues/1159).  Storing intervals instead of `positional_metadata` would substantially save on memory.
> > +
> > +Additional benefits of having a `interval_metadata` objects include the ability to rapidly query features by interval.  IntervalTrees allow for both querying by position and ranges to retrieve all overlapping intervals.  This can be particularly helpful when dealing with layers of features (i.e. transcripts made up of exons, or operons made up of genes).
> > +
> > +# Detailed design
> > +
> > +We propose to use the `IntervalTree` object in `bx-python.`  Some benchmarks have been made for various IntervalTree data structures - in this [benchmark](https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was determined to have one of the fastest query times.
> > +
> > +## `IntervalMetadata` object structure
> > +features : dictionary.  keys are `skbio.Feature` objects.  values are a list of tuples, where each tuple corresponds to an interval.
> > +intervals : bx-python `IntervalTree` object.  The keys correspond to intervals and the values correspond to a single `skbio.Feature` object.
> > +
> > +These objects are arranged such that features can be queried by both feature attributes and intervals.
> > +
> > +## `Feature` object structure
> > +The feature object is more a less a immutable, hashable dictionary.  This store arbiturary information about features.  For instance, strand information, CDS, ...
>
> Daniel, you are looking at obsoleted version
> … <#m_-3759575383248016830_>
> On Fri, Apr 29, 2016 at 11:05 AM, Daniel McDonald ***@***.***> wrote: In
> active/003-seq-feature.md <#9 (comment)
> <https://github.com/biocore/scikit-bio-rfcs/pull/9#discussion_r61620972>>:
> > +The current implementation of `positional_metadata` in the `Sequence`
> class takes > 1TB memory to read in a E. coli genome with 260K features.
> And the `pandas` sparse data frame can alleviate the mem usage, but it is
> still too memory hogging. The discussion is mainly in this [thread](
> biocore/scikit-bio#1159
> <https://github.com/biocore/scikit-bio/issues/1159>). Storing intervals
> instead of `positional_metadata` would substantially save on memory. > + >
> +Additional benefits of having a `interval_metadata` objects include the
> ability to rapidly query features by interval. IntervalTrees allow for both
> querying by position and ranges to retrieve all overlapping intervals. This
> can be particularly helpful when dealing with layers of features (i.e.
> transcripts made up of exons, or operons made up of genes). > + > +#
> Detailed design > + > +We propose to use the `IntervalTree` object in
> `bx-python.` Some benchmarks have been made for various IntervalTree data
> structures - in this [benchmark](
> https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was
> determined to have one of the fastest query times. > + > +##
> `IntervalMetadata` object structure > +features : dictionary. keys are
> `skbio.Feature` objects. values are a list of tuples, where each tuple
> corresponds to an interval. > +intervals : bx-python `IntervalTree` object.
> The keys correspond to intervals and the values correspond to a single
> `skbio.Feature` object. > + > +These objects are arranged such that
> features can be queried by both feature attributes and intervals. > + > +##
> `Feature` object structure > +The feature object is more a less a
> immutable, hashable dictionary. This store arbiturary information about
> features. For instance, strand information, CDS, ... Thanks. Would it be
> possible to update the RFC? The ambiguity of "...is more a less a
> immutable..." is what I'm specifically hoping to resolve — You are
> receiving this because you were mentioned. Reply to this email directly or
> view it on GitHub <
> https://github.com/biocore/scikit-bio-rfcs/pull/9/files/b954b92abe3eb629826cd2a898461e97e57dd057#r61620972>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio-rfcs/pull/9/files/b954b92abe3eb629826cd2a898461e97e57dd057#r61621223>
>